### PR TITLE
Search socket dir before falling back to psycopg2 guess

### PR DESCRIPTION
**Pull request by bersace**
Closes: #37 

As suggested in the description of the issue, I just use the first existing socketdir found from a list of well known candidates. This should be enough.

### CODE REVIEW

File: pgcli/main.py [hunk 1]
```diff
@@ -316,9 +316,23 @@ def fixup_possible_percent_encoding(s):
...
+
+        for candidate in candidates:
+            if os.path.isdir(candidate):
+                return candidate
+
     def connect(self, database='', host='', user='', port='', passwd='',
                 dsn='', **kwargs):
         # Connect to the database.
+        if not host:
+            host = self.guess_socketdir() or ''
```
Comment (bersace) [inline pgcli/main.py]: Actually, I do not like mixing param guess inside `connect` function. This is misleading. I suggest writing a `guess_dsn` function that do all the guess I/O, env var fallback, etc. and let `connect` just trigger the connection or fail.

Comment (j-bennet) [inline pgcli/main.py]: @bersace Feel free to extract the "guesswork" into a separate method. You're right that the `connect` method is becoming difficult to read.

Comment (bersace) [inline pgcli/main.py]: I prefer working on this in another PR :-)

**Comment by bersace** [comment]
@j-bennet Ready for merge.

**Comment by bersace** [comment]
🆙 @j-bennet

**Comment by amjith** [comment]
@bersace This change made my connection experience worse. I connect to my local instance via localhost:5432. But this change give preference to socket over port. 

Previously I used to type `pgcli` and nothing else it will correctly connect to postgres. Now if I type `pgcli` I get the following error: 

```
could not connect to server: No such file or directory
        Is the server running locally and accepting
        connections on Unix domain socket "/usr/local/var/postgres/.s.PGSQL.5432"?
```

**Comment by bersace** [comment]
@amjith That's not the goal ^^

What is the behaviour of psql for you ? Do you want `pgcli` to glob for `.PGSQL*` socket ?

For the sharing, I run simple docker compose file and use `direnv` to setup `PGHOST` when I cd to `pgcli` directory.

``` yaml
# cat /my-docker-compose.yml
version: '2'

services:
  pg96:
    image: postgres:9.6
```

``` bash
# cat .envrc
export PGHOST=postgres.docker
export PGUSER=postgres
. ${WORKON_HOME}/pgcli/bin/activate
export COMPOSE_FILE=my-docker-compose.yml
```

**Comment by amjith** [comment]
Right now the behavior pgcli mimics the behavior of psql. I don't think there is a standard name for postgres socket files.

I'm inclined to say that we stick with psycopg2's order of detection to connect to a database and not do anything fancy ourselves.

**Comment by j-bennet** [comment]
I think this PR was caused by https://github.com/dbcli/pgcli/issues/37, where `pgcli` failed to connect to a socket, while `psql` did the right thing. It would be good to fix that. I think it should be possible to check for being able to connect on those sockets first, and if that fails, delegate to psycopg2 to do the rest.

**Comment by j-bennet** [comment]
So I verified the current `pgcli` vs `psql` behavior, and it seems to me that pgcli actually does the right thing. It can connect to the UNIX socket if it is correctly specified in `postgresql.conf` as `unix_socket_directory` or `unix_socket_directories`. If it's not correctly specified there, then `psql` would not be able to connect either (see https://dba.stackexchange.com/questions/21587/postgresql-is-running-locally-but-i-cannot-connect-why).

I think this change is not needed, sorry @bersace.

**Comment by amjith** [comment]
Closing this in favor of existing behavior.

**Pull request closed without merging**